### DISCUSSION TITLE
How to get xref associated with a text span

### DISCUSSION THREAD

**Discussion by SirishaGorasa** [pymupdf/PyMuPDF#1256] [Q&A]
How do I get the xref associated with a text span?

I have seen in previous discussions, as we can't extract headings directly using pymupdf. Wanna know if there are any improvements!

**Comment by JorjMcKie** [comment]
Text components are not associated with xrefs.
Headers, footers, etc. are no identifyable information in PDF - has nothing to do with PyMuPDF. Neither are text blocks or spans.
The base library MuPDF does its bests to use a bunch of heuristics to identify blocks, lines and spans - but there is no materal foundation in the document for such interpretation.
You may for example encounter PDFs where all the text characters are coded in an arbitrary sequence - obviously with up to ``n!`` different possibilities of extraction sequences if n is the number of characters on a page. I have created such an awkward example myself: https://github.com/pymupdf/PyMuPDF-Utilities/blob/master/text-extraction/textmaker2.pdf.
Similarly it is very possible to extract a header line as the last text piece from a page.
So you need to use text positioning coordinates to establish a reading sequence you like. This can also not be generalized - think of non-Western reading sequences (right to left: Arabian, Hebrew) or top-down-left-right for some ancient Chinese or Japanese documents.

But there is help: use ``python -m fitz gettext textmaker2.pdf`` to extract and re-layout all the single characters in the intended reading sequence.

**Reply by SirishaGorasa** [reply]
So, how do headings are tagged with their levels for example in Adobe?

**Reply by JorjMcKie** [reply]
they are not

**Reply by JorjMcKie** [reply]
In PDF there is no such thing like a "heading". **You** interpret a piece of text as such, based on it page position and maybe other criteria like font properties.

**Reply by SirishaGorasa** [reply]
Yes, But we can see tag tree in Adobe for PDFs which include information related to headings, links, paragraphs etc. Are you aware of this?

**Reply by JorjMcKie** [reply]
I am not sure what you mean: example?

**Reply by SirishaGorasa** [reply]
Similar to what we have in HTML

**Reply by JorjMcKie** [reply]
If you mean tagged PDFs in the sense that this technique is used for structuring content, then no, this is not supported in MuPDF.

**Reply by SirishaGorasa** [reply]
Do you have any references to making such structured contents?

**Reply by JorjMcKie** [reply]
There also are ways to formally store tabular data (tables) in a PDF, making cells, columns, rows identifyable.
This too is not supported.

**Reply by JorjMcKie** [reply]
I am only aware of the original PDF specs https://www.adobe.com/content/dam/acom/en/devnet/acrobat/pdfs/pdf_reference_1-7.pdf, https://www.adobe.com/content/dam/acom/en/devnet/pdf/pdfs/PDF32000_2008.pdf.

Once you have understood how these things must be "coded", you can indeed use PyMuPDF's low-level coing support to store them in PDF objects.
Vice versa: if you know how to access such information, you can use low-level code to do it.
But I have nothing prepared or in preparation to make this easy.
I might if somebody pays me a fortune 😎

The problem here is like always, that I don't want to extend MuPDF too much with additional code. This increases my work load with additional probability of introducing bugs, and, second, MuPDF might decide to support something of that, which would shatter my solution in pieces.

**Reply by SirishaGorasa** [reply]
Ok @JorjMcKie, got it. Could you please help me with a small example of how to use mupdf's low level coding to do such thing?

**Reply by JorjMcKie** [reply]
You can use PyMuPDF to
* create new PDF objects togther with a new xref
```python
xref = doc.get_new_xref()
doc.update_object(xref, "<< ... object definition >>")
```
* read the source of object definitions and associated streams
```python
obj_source = doc.xref_object(xref) # a string
stream = doc.xref_stream(xref) # a bytes object
```
* read / update / edit single value in an object
```python
resources = doc.xref_get_key(page.xref, "Resources") # read the /Resources definition of a page definition
... ect.
```
Also see the documentation https://pymupdf.readthedocs.io/en/latest/faq.html#low-level-interfaces

**Reply by buptyyf** [reply]
Is there a way to get bbox or position info of a xref? I use doc. xref_object(xref) only get a string as follow:
```
<<\n  /A 80 0 R\n  /Next 76 0 R\n  /Parent 40 0 R\n  /Title (2 | Flexible Scaling for Analytic Workloads)\n>>
```
This string maybe includes text and parent xref etc. But I don't know the introduction of each character. And this maybe don't include bbox info.

**Reply by JorjMcKie** [reply]
> Is there a way to get bbox or position info of a xref? I use doc. xref_object(xref) only get a string as follow:
> 
> ```
> <<\n  /A 80 0 R\n  /Next 76 0 R\n  /Parent 40 0 R\n  /Title (2 | Flexible Scaling for Analytic Workloads)\n>>
> ```
> 
> This string maybe includes text and parent xref etc. But I don't know the introduction of each character. And this maybe don't include bbox info.

This xref looks like a PDF bookmark. This is not text or whatever, so does not have a bbox. It does **_not appear_** anywhere on the page.

**Reply by buptyyf** [reply]
doc.get_toc() sometimes only returns xref info and don't have point info. I just want to know where the toc point to, but I can't. Are there some alternative ways let me know that?

**Reply by JorjMcKie** [reply]
> doc.get_toc() sometimes only returns xref info and don't have point info. I just want to know where the toc point to, but I can't. Are there some alternative ways let me know that?

This happens if the creator did not specify a target point. Did you confirm that that bookmark points to some place inside the same PDF - check the "kind" value. Also if the page is < 1 means that the target is not within this PDF.

**Reply by buptyyf** [reply]
Example PDF:
[CB_7YGBwmBxZEhw6eK6cN4Ot6eb.pdf](https://github.com/pymupdf/PyMuPDF/files/10755614/CB_7YGBwmBxZEhw6eK6cN4Ot6eb.pdf)

doc.get_toc(False) get as follows:
```
[1, '上篇', 7, {'kind': 4, 'xref': 1033, 'name': 'page=7&zoom=nan,90,117', 'collapse': False, 'zoom': 0.0}]
[2, '第一章 实际GDP是假的', 7, {'kind': 4, 'xref': 1101, 'name': 'page=7&zoom=nan,268,171', 'collapse': False, 'zoom': 0.0}]
...
```
It's a named location toc. But I don't it's actual position.

**Reply by JorjMcKie** [reply]
You are not using the current version. There was an error before that did not recognize "&zoom=nan" as being equivalent to "&zoom=0".
Upgrading takes care of this, so that these links are normal internal GOTO links.
